### PR TITLE
reliability improvements

### PR DESCRIPTION
**Pull request by SreeramGarlapati**
1. Transient errors should bubble-up to client after retryCount limit exceeds 
2. Fix a bug in code path: ConnectionCreation should be lazy - triggered by actual operation
****sender testing in progress. recv'r already tested.

### CODE REVIEW

**Comment by SreeramGarlapati** [comment]
This will improve the user experience in the issue: #179

File: java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java [hunk 1]
```diff
@@ -300,16 +300,16 @@ public void setReceiveTimeout(final Duration value)
...
 					final List<Message> messages = receiveCore(maxMessageCount);
 					if (messages != null)
 						onReceive.complete(messages);
 					else
 						pendingReceives.offer(new ReceiveWorkItem(onReceive, receiveTimeout, maxMessageCount));
+
+					if (receiveLink.getLocalState() == EndpointState.CLOSED || receiveLink.getRemoteState() == EndpointState.CLOSED)
+					{
+						createReceiveLink();
+					}
```
Comment (hmlam) [inline java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java]: If this order of "put workItem into pendingQueue" --> "recreate Link if broken" actually matters in terms of race condition, I think we should have some base implementation that enforce this order everywhere so that sub-class implementation does not make this type of mistake easily (e.g. I am worry when we write the messaging implementation we would hit the same thing again).

**Comment by hmlam** [comment]
:clock1:

File: java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionSender.java [hunk 2]
```diff
@@ -61,7 +61,8 @@ public PartitionSender apply(Void a)
 	/**
 	 * Synchronous version of {@link #send(EventData)} Api. 
 	 * @param data the {@link EventData} to be sent.
-	 * @throws ServiceBusException if Service Bus service encountered problems during the operation.
+	 * @throws PayloadSizeExceededException    if the total size of the {@link EventData} exceeds a pre-defined limit set by the service. Default is 256k bytes.
```
Comment (SreeramGarlapati) [inline java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionSender.java]: these Javadoc fixes are unrelated to this.. but included them as they are small & harmless.

File: java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java (hunk 1 above)
Comment (SreeramGarlapati) [inline java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java]: good comment @hmlam. I belv. a strict contract needs to be incorporated in couple other places. filed an issue to track this: #244
